aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -254,6 +255,249 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
   return at::linalg_matrix_rank(self, c10::nullopt, symmetric);
 }
 
+// multi_dot helper functions
+namespace {
+
+/**
+ * @brief Computes the optimal order to multiply the input tensors
+ *
+ * Follows the dynamic programming algorithm from Cormen et al,
+ * "Introduction to Algorithms, Third Edition", Chapter 15.2,
+ * p. 370-378. Note that the book uses 1-based indexing.
+ *
+ * The cost of multiplying two matrices with sizes p x q and q x r
+ * is defined here as p * q * r. The optimal multiplication order
+ * is the one that minimizes the total cost.
+ *
+ * @param tensors list of 2D tensors
+ * @return 2D list used by #matrix_chain_multiplication
+ */
+std::vector<std::vector<int64_t>> matrix_chain_order(
```

Comment:
why do we have this copypaste from chain_matmul?

Reply:
Because I didn't know about the chain_matmul operator before I wrote this PR and both follow the CLRS book. I think the plan now is to deprecate chain_matmul in favor of this one for NumPy compatibiltiy.

Reply:
Even if the plan is to deprecate chain_matmul (why?) chain_matmul and this should share implementation.

Reply:
Yes, I was going to make chain_matmul call into multi_dot in a separate PR that also adds the deprecation note in the docs. @mruberry What's your opinion on whether we should deprecate it or just have it call into multi_dot to share the code?

Reply:
One impl in this PR, aliasing/deprecation/docs updates for chain_matmul in a separate PR.